tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+
+        private const string Title = "Test classes should be derived from BaseTestClass";
+        private const string Format = "Test class '{0}' should extend BaseTestClass.";
+        private const string Description =
+            "Test classes should be derived from BaseTestClass.";
+
+        private static DiagnosticDescriptor Rule =
+            new DiagnosticDescriptor(DiagnosticId, Title, Format, Category,
+                DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(AnalyzeCompilation);
+        }
+
+        private void AnalyzeCompilation(CompilationStartAnalysisContext context)
+        {
+            if (!(context.Compilation.GetTypeByMetadataName("Xunit.FactAttribute") is { } factAttribute))
+            {
+                return;
+            }
+
+            var analyzerImpl = new AnalyzerImpl(context.Compilation, factAttribute);
+            context.RegisterSymbolAction(analyzerImpl.AnalyzeNamedType, SymbolKind.NamedType);
+        }
+
+        private sealed class AnalyzerImpl
+        {
+            private readonly Compilation _compilation;
+            private readonly INamedTypeSymbol _factAttribute;
+            private readonly INamedTypeSymbol _baseTestClass;
+            private readonly ConcurrentDictionary<INamedTypeSymbol, bool> _knownTestAttributes = new ConcurrentDictionary<INamedTypeSymbol, bool>();
+
+            public AnalyzerImpl(Compilation compilation, INamedTypeSymbol factAttribute)
+            {
+                _compilation = compilation;
+                _factAttribute = factAttribute;
+                _baseTestClass = _compilation.GetTypeByMetadataName("Microsoft.ML.TestFramework.BaseTestClass");
+            }
+
+            public void AnalyzeNamedType(SymbolAnalysisContext context)
+            {
+                var namedType = (INamedTypeSymbol)context.Symbol;
+                if (namedType.TypeKind != TypeKind.Class)
+                    return;
+
+                if (ExtendsBaseTestClass(namedType))
+                    return;
+
+                var hasTestMethod = false;
+                foreach (var member in namedType.GetMembers())
+                {
+                    if (member is IMethodSymbol method && IsTestMethod(method))
+                    {
+                        hasTestMethod = true;
+                        break;
+                    }
+                }
+
+                if (!hasTestMethod)
+                    return;
+
+                context.ReportDiagnostic(Diagnostic.Create(Rule, namedType.Locations[0], namedType));
```

Comment:
At this point, we have a class with test methods, but the class did not extend `BaseTestClass`. We report a diagnostic instance, using the descriptor `Rule` defined above. The location is the identifier for the class, and the third parameter (`namedType`) is the message argument for the `{0}` in the message format.